server/src/main/java/org/opensearch/index/remote/RemoteRefreshSegmentTracker.java

```diff
@@ -0,0 +1,425 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index.remote;
+
+import org.opensearch.common.util.MovingAverage;
+import org.opensearch.common.util.Streak;
+import org.opensearch.common.util.concurrent.ConcurrentCollections;
+import org.opensearch.index.shard.ShardId;
+
+import java.util.Map;
+import java.util.Set;
+import java.util.concurrent.atomic.AtomicLong;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
+
+/**
+ * Keeps track of remote refresh which happens in {@link org.opensearch.index.shard.RemoteStoreRefreshListener}. This consist of multiple critical metrics.
+ */
+public class RemoteRefreshSegmentTracker {
+
+    public RemoteRefreshSegmentTracker(
+        ShardId shardId,
+        int uploadBytesMovingAverageWindowSize,
+        int uploadBytesPerSecMovingAverageWindowSize,
+        int uploadTimeMsMovingAverageWindowSize
+    ) {
+        this.shardId = shardId;
+        // Both the local refresh time and remote refresh time are set with current time to give consistent view of time lag when it arises.
+        long currentTimeMs = System.nanoTime() / 1_000_000L;
+        localRefreshTimeMs.set(currentTimeMs);
+        remoteRefreshTimeMs.set(currentTimeMs);
+        this.uploadBytesMovingAverageReference = new AtomicReference<>(new MovingAverage(uploadBytesMovingAverageWindowSize));
+        this.uploadBytesPerSecMovingAverageReference = new AtomicReference<>(new MovingAverage(uploadBytesPerSecMovingAverageWindowSize));
+        this.uploadTimeMsMovingAverageReference = new AtomicReference<>(new MovingAverage(uploadTimeMsMovingAverageWindowSize));
+    }
+
+    /**
+     * ShardId for which this instance tracks the remote segment upload metadata.
+     */
+    private final ShardId shardId;
+
+    /**
+     * Every refresh is assigned a sequence number. This is the sequence number of the most recent refresh.
+     */
+    private final AtomicLong localRefreshSeqNo = new AtomicLong();
```

Comment:
I spot checked a few, but I only see usages of `get` and `set`. Do you need AtomicLongs as opposed to just using a volatile primitive `long`? The Atomic* variants are generally only needed if you're doing things like compareAndSet or updateAndGet.

Reply:
Had kept it AtomicLong for thread safety in mind. At the moment we do not have concurrent usecases, will see if we can change this to volatile long.